### DISCUSSION TITLE
notebook breaks in Pluto 0.18.0 while works in 0.17.3

### DISCUSSION THREAD

**Discussion by AbhimanyuAryan** [fonsp/Pluto.jl#1910] [Questions about using Pluto] [answered]
Packages install using PlutoPkg are not found `SearchLight` `SearchLightSQLite`

Video on notebook working in Pluto 0.17.3 (works)

https://user-images.githubusercontent.com/8083613/153439115-850a5caa-3868-471e-8e35-bae659ef9d56.mp4

Video on notebook breaking in Pluto 0.18.0 (breaks)

https://user-images.githubusercontent.com/8083613/153439134-fce41117-7beb-467e-8f50-7a72a60d2000.mp4

📝 Pluto notebook file: https://gist.github.com/AbhimanyuAryan/cb603f44c4186eeb5bf310204a81a496 (it is little long that was best I could do to strip it down. All components are dependent on each other)

Start breaking  near this [line](https://gist.github.com/AbhimanyuAryan/cb603f44c4186eeb5bf310204a81a496#file-pluto_bug-jl-L645) where `SearchLight` is being used

**Comment by fonsp** [comment, marked as answer]
> Packages install using PlutoPkg are not found `SearchLight` `SearchLightSQLite`

What does this mean?

Thanks for providing the video and the notebook file, but what is the problem exactly? 
- *What* is broken? 
- How can you tell? 
- Why should it not be broken? 
- What have you tried already?

**Reply by AbhimanyuAryan** [reply]
* What is broken?
  - Fons if you see both videos. I'm importing `using SearchLight` & `using SearchLightSQLite` in 2 cells. Which I am using in below cells but they somehow are not found/installed in the notebook. Strange now this exact notebook doesn't break in 0.17.3 but it's breaking with Pluto 0.18.0

* How can I tell?
  - Same notebook works with Pluto 0.17.3 I gave you videos of both. All cells which uses `SearchLight` & `SearchLightSQLite` will throw errors: https://gist.github.com/AbhimanyuAryan/cb603f44c4186eeb5bf310204a81a496#file-pluto_bug-jl-L637-L679 but only with Pluto 0.18.0 not with 0.17.3

* Why should it not be broken?
  - Why does it work with 0.17.3? Because I'm installing `SearchLight` and `SearchLightSQLite` with Pluto's package manager. Technically, it should allow me to use those packages right? 
  - Also, wrapping these packages with dummy variable seems to solve the problem

* What have you tried already?
  ```julia
   begin
	    db;
   	using SearchLightSQLite;
   end
   ```

but once the `db dummy variable` solves the problem of `SearchLight` and `SeachLightSQLite` then `using SQLite` breaks. Is it that because `SQLite` is already a dependency for `SearchLightSQLite` and that is already installed so doing `using SQLite` again causes the problem?

**Reply by AbhimanyuAryan** [reply]
i.e. 

```julia
UndefVarError: SQLite not defined

    top-level scope@[Local: 5](http://localhost:1234/edit?id=627c72e2-8e31-11ec-2774-ebfe561f3ee3#)
```

**Reply by AbhimanyuAryan** [reply]
<img width="1016" alt="Screenshot 2022-02-15 at 1 13 35 PM" src="https://user-images.githubusercontent.com/8083613/154015518-701d7ef3-2383-46f1-bc76-636356345d99.png">

**Reply by AbhimanyuAryan** [reply]
Ok this is fixed. It think it was because of double import. `SearchLightSQLite.SQLite.DB` worked

**Comment by fonsp** [comment]
It looks like the notebook is relying on the order of side effects. For example, these three cells:
```julia
filename = "hello"
```
```julia
write(filename, "world!")
```
```julia
read(filename, String) == "world!"
```

are not guaranteed to run top to bottom, because we run cells in "topological order" instead. Pluto might run this notebook in the order 1, 3, 2, because this respects all variable links. 

You can fix this by creating dummy variables, like you are already doing in some places:
```julia
filename = "hello"
```
```julia
written = write(filename, "world!")
```
```julia
written; read(filename, String) == "world!"
```

Note that now there is an explicit link `2 -> 3`, meaning that *2 must run before 3*. Just referencing `filename` in the third cell is not enough.

An example in your notebook:
Cell A: https://gist.github.com/AbhimanyuAryan/cb603f44c4186eeb5bf310204a81a496#file-pluto_bug-jl-L660
Cell B: https://gist.github.com/AbhimanyuAryan/cb603f44c4186eeb5bf310204a81a496#file-pluto_bug-jl-L685

Because of side effects (writing to the `/migrations` folder), B must run *after* A, but there is no variable linking the two. They both depend on `db`, but this does not define their relative order. 

There are some more hidden dependencies between cells in the notebook, like these four cells: https://gist.github.com/AbhimanyuAryan/cb603f44c4186eeb5bf310204a81a496#file-pluto_bug-jl-L608-L640 I recommend putting these into a single cell.

**Reply by AbhimanyuAryan** [reply]
Thanks fons for these suggestions. I'll refactor these guides/notebooks and open source them soon which will serve as a reference for anyone interested in "How to use `Dummy Variable`/`explicit execution`/`file operations` with Pluto". Again thanks a lot of detailed reply

**Reply by AbhimanyuAryan** [reply]
also thanks for providing `dummy variables` for explicit execution with Pluto. They are were neat
